File: sdk/python/agenthub/memory.py
"""Memory configuration for AgentBreeder agents.

Supports in-memory, PostgreSQL, and Redis backends with configurable
buffer strategies.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class MemoryConfig:
    """Serializable memory configuration."""

    backend: str = "in_memory"
    memory_type: str = "buffer_window"
    max_messages: int = 100
    namespace_pattern: str = "{agent_id}:{session_id}"

    def to_dict(self) -> dict:
        """Serialize to a dict suitable for YAML output."""
        d: dict = {"backend": self.backend, "memory_type": self.memory_type}
        if self.max_messages != 100:
            d["max_messages"] = self.max_messages
        if self.namespace_pattern != "{agent_id}:{session_id}":
            d["namespace_pattern"] = self.namespace_pattern
        return d


class Memory:
    """Configure memory backends programmatically."""

    def __init__(self, config: MemoryConfig | None = None) -> None:
        self.config = config or MemoryConfig()

    @staticmethod
    def buffer_window(max_messages: int = 100, **kwargs) -> Memory:
        """Create a buffer-window memory (sliding window of recent messages)."""
        return Memory(
            MemoryConfig(
                memory_type="buffer_window",
                max_messages=max_messages,
                **kwargs,
            )
        )

    @staticmethod
    def buffer(**kwargs) -> Memory:
        """Create an unbounded buffer memory."""
        return Memory(MemoryConfig(memory_type="buffer", **kwargs))

    @staticmethod
    def postgresql(**kwargs) -> Memory:
        """Create a PostgreSQL-backed memory."""
        return Memory(MemoryConfig(backend="postgresql", **kwargs))

    def to_config(self) -> MemoryConfig:
        """Return the underlying MemoryConfig."""
        return self.config

    def to_dict(self) -> dict:
        """Serialize to a dict suitable for YAML output."""
        return self.config.to_dict()

File: sdk/typescript/src/memory.ts
/** Memory configuration for AgentBreeder agents. */

export interface MemoryConfig {
  backend: string;
  maxMessages?: number;
  connectionString?: string;
}

export class Memory {
  private _config: MemoryConfig;

  private constructor(config: MemoryConfig) {
    this._config = config;
  }

  static bufferWindow(maxMessages?: number, opts?: Partial<MemoryConfig>): Memory {
    return new Memory({ backend: "buffer_window", maxMessages: maxMessages ?? 10, ...opts });
  }

  static buffer(opts?: Partial<MemoryConfig>): Memory {
    return new Memory({ backend: "buffer", ...opts });
  }

  static postgresql(opts?: Partial<MemoryConfig>): Memory {
    return new Memory({ backend: "postgresql", ...opts });
  }

  static fromConfig(config: MemoryConfig): Memory {
    return new Memory({ ...config });
  }

  toConfig(): MemoryConfig {
    return { ...this._config };
  }
}
